smac_planner/src/a_star.cpp

```diff
@@ -364,6 +398,97 @@ bool AStarAlgorithm<NodeT>::isGoal(NodePtr & node)
   return node == getGoal();
 }
 
+template<>
+AStarAlgorithm<NodeSE2>::NodePtr AStarAlgorithm<NodeSE2>::getAnalyticPath(
+  const NodePtr & node,
+  const NodeGetter & node_getter)
+{
+  std::vector<std::pair<NodePtr, Coordinates>> possible_nodes;
+  const NodePtr & goal = getGoal();
+  ompl::base::ScopedState<> from(node->motion_table.state_space), to(
+    node->motion_table.state_space), s(node->motion_table.state_space);
+  const NodeSE2::Coordinates & node_coords = node->pose;
+  const NodeSE2::Coordinates & goal_coords = goal->pose;
+  from[0] = node_coords.x;
+  from[1] = node_coords.y;
+  from[2] = node_coords.theta * node->motion_table.bin_size;
+  to[0] = goal_coords.x;
+  to[1] = goal_coords.y;
+  to[2] = goal_coords.theta * node->motion_table.bin_size;
+
+  float d = node->motion_table.state_space->distance(from(), to());
+  NodePtr prev(node);
+  // A move of sqrt(2) is guaranteed to be in a new cell
+  constexpr float sqrt_2 = std::sqrt(2.);
+  unsigned int num_pts = std::floor(d / sqrt_2) + 1;
+  std::vector<double> reals;
+  // Pre-allocate
+  unsigned int index = 0;
+  NodePtr next(nullptr);
+  float angle = 0.0;
+  Coordinates proposed_coordinates;
+  // Don't generate the first point because we are already there!
+  // And the last point is the goal, so ignore it too!
+  for (unsigned int i = 1; i < num_pts; i++) {
+    node->motion_table.state_space->interpolate(from(), to(), (double)i / num_pts, s());
+    reals = s.reals();
+    angle = reals[2] / node->motion_table.bin_size;
+    while (angle >= node->motion_table.num_angle_quantization_float) {
+      angle -= node->motion_table.num_angle_quantization_float;
+    }
+    while (angle < 0.0) {
+      angle += node->motion_table.num_angle_quantization_float;
+    }
+    // Turn the pose into a node, and check if it is valid
+    index = NodeSE2::getIndex(
+      static_cast<unsigned int>(reals[0]),
+      static_cast<unsigned int>(reals[1]),
+      static_cast<unsigned int>(angle));
+    // Get the node from the graph
+    if (node_getter(index, next)) {
+      Coordinates initial_node_coords = next->pose;
+      proposed_coordinates = {static_cast<float>(reals[0]), static_cast<float>(reals[1]), angle};
+      next->setPose(proposed_coordinates);
+      if (next->isNodeValid(_traverse_unknown) && next != prev) {
+        // Set coordinates
+        possible_nodes.emplace_back(next, proposed_coordinates);
+        prev = next;
+      } else {
+        next->setPose(initial_node_coords);
+        return NodePtr(nullptr);
+      }
+    } else {
+      // Abort
+      return NodePtr(nullptr);
+    }
+  }
+  if (num_pts > 2) {
+    // Legitimate path - set the parent relationships and poses
+    NodePtr prev = node;
+    for (const auto & node_pose : possible_nodes) {
+      const auto & n = node_pose.first;
+      if (n != prev) {
+        n->parent = prev;
+        n->setPose(node_pose.second);
```

Comment:
So you've already set the pose above in L451 with the proposed coordinates. An issue that can arise is that if we search for the analytic expansion but don't find it, we just overwrote all the poses and that will mess things up with continuity. 

I think there are 2 options here. Do what you did in the main for loop, but then if its in collision or invalid, we need to enter another for loop to reset with the initial poses. You stored the `proposed_coordinates` here which were already set, but I think we need to invert it and store the initial coordinates to reset if `num_pts < 2` or there was a collision. 

Right now if there's a collision you return a nullptr but you don't actually reverse all the previously searched nodes that we've now potentially changed the coordinates of

Reply:
Good point - oversight on my end.
Fixed.